TheaterSchedule.BLL/Services/ResetCodeService.cs

```diff
@@ -0,0 +1,106 @@
+﻿using System;
+using TheaterSchedule.BLL.Helpers;
+using TheaterSchedule.BLL.DTOs;
+using TheaterSchedule.BLL.Interfaces;
+using TheaterSchedule.DAL.Interfaces;
+using System.Linq;
+using Microsoft.EntityFrameworkCore;
+using TheaterSchedule.DAL.Models;
+using System.Threading.Tasks;
+using System.Net;
+using TheaterSchedule.Infrastructure;
+
+namespace TheaterSchedule.BLL.Services
+{
+    public class ResetCodeService : IResetCodeService
+    {
+        private IResetCodeRepository _resetCodeRepository;
+        private ITheaterScheduleUnitOfWork _theaterScheduleUnitOfWork;
+        private IUserRepository _userRepository;
+        private IEmailService _emailService;
+
+        public ResetCodeService(ITheaterScheduleUnitOfWork theaterScheduleUnitOfWork, IResetCodeRepository resetCodeRepository, IUserRepository userRepository, IEmailService emailService)
+        {
+            _theaterScheduleUnitOfWork = theaterScheduleUnitOfWork;
+            _resetCodeRepository = resetCodeRepository;
+            _userRepository = userRepository;
+            _emailService = emailService;
+        }
+
+        public async Task<ResetCodeDTO> GetByValueAsync(int value)
+        {
+            var resetCode = await _resetCodeRepository.GetByValueAsync(value);
+
+            if (resetCode == null)
+            {
+                throw new HttpStatusCodeException(HttpStatusCode.NotFound, $"Such rest code doesn't exists");
+            }
+
+            return new ResetCodeDTO
+            {
+                Id = resetCode.Id,
+                Code = resetCode.Code
+            };
+        }
+
+        public void GenerateResetCodeAsync(string email)
+        {
+            var user = _resetCodeRepository.GetUserByEmailAddress(email);
+
+            if (user == null)
+            {
+                throw new HttpStatusCodeException(HttpStatusCode.NotFound, $"Such user doesn't exist");
+            }
+
+            int random = new Random().Next(1000, 9999);
+            TimeSpan currentTime = new TimeSpan(DateTime.Now.Hour, DateTime.Now.Minute, DateTime.Now.Second);
+
+            _resetCodeRepository.Add(new ResetCodeModel
+            {
+                Code = random,
+                AccountId = user.Result.AccountId,
+                CreationTime = currentTime
+            });
+
+            _emailService.SendEmailAsync(email, "noreply", $"Verification code for resetting your password: {random}");
+
+            _theaterScheduleUnitOfWork.Save();
+        }
+
+        public void ValidateCode(int validationCode)
+        {
+            var code = _resetCodeRepository.GetByValueAsync(validationCode);
+
+            if (code == null)
+            {
+                throw new HttpStatusCodeException(HttpStatusCode.NotFound);
+            }
+
+            if ((DateTime.Now.Minute - code.Result.CreationTime.Minutes) > 5)
```

Comment:
what if user will try exactly in 5 hours? Minutes will be the same
Subtract dates  - it will return TimeSpan object which represents difference beetween time
and then use minutes from timespan

Reply:
Changed to TimeSpan.Substract  - because can not convert TimeSpan to DateTime.
Functionality almost the same as in DateTime.Substract.